src/Nethermind/Nethermind.Blockchain/Receipts/PersistentReceiptStorage.cs

```diff
@@ -358,14 +359,16 @@ public void EnsureCanonical(Block block)
             {
                 foreach (Transaction tx in block.Transactions)
                 {
-                    writeBatch[tx.Hash.Bytes] = Rlp.Encode(block.Number).Bytes;
+                    Hash256 hash = (tx.Hash ??= tx.CalculateHash());
+                    writeBatch[hash.Bytes] = Rlp.Encode(block.Number).Bytes;
```

Comment:
we should do this once before the loop, also we can probably use PutSpan to avoid allocating

Reply:
Moved. Not sure the WriteBatch supports PutSpan